certbot/certbot/_internal/renewal.py

```diff
@@ -325,13 +381,26 @@ def renew_cert(config: configuration.NamespaceConfig, domains: Optional[List[str
     _avoid_invalidating_lineage(config, lineage, original_server)
     if not domains:
         domains = lineage.names()
-    # The private key is the existing lineage private key if reuse_key is set.
-    # Otherwise, generate a fresh private key by passing None.
-    if config.reuse_key:
-        new_key = os.path.normpath(lineage.privkey)
-        _update_renewal_params_from_key(new_key, config)
+
+    # Determine private key action.
+    # - Key param change requested on the CLI?
+    #   - If reuse_key, make sure user confirms with --cert-name
+    #   - Otherwise, apply the change.
+    # - Key params on the CLI unchanged?
+    #   - If reuse_key, use the existing private key
+    #   - Otherwise, generate a new private key using key params of existing key.
+    is_key_change_requested = _handle_reuse_key_conflict(config, lineage)
+
+    existing_key = os.path.normpath(lineage.privkey)
+    new_key: Optional[str] = None
+
+    if is_key_change_requested:
```

Comment:
What's the scenario where we need to update renewal params from the key when `reuse_key` is not set? The only scenario that immediately comes to mind for me is `certbot renew` with a manually modified renewal configuration file.

My thinking here is very early in Certbot's execution for any subcommand, a config object is created through a combination of CLI params, global config file, and CLI defaults. If the user ran `certonly` or `run`, I think this combination of factors should already contain the key options the user wants. For `certbot renew`, we've already combined the config I just described with the renewal parameters in `_reconstitute` which is what I think is correct in this case.

My initial impression is we shouldn't worry about the case of a manually modified renewal config file, but please let me know if you think otherwise.

Unless I'm missing something, I think this code can be simplified by:

1. Removing `_update_renewal_params_from_key_and_config` and any calls to it.
2. Only calling `_update_renewal_params_from_key` and `_handle_reuse_key_conflict` when `config.reuse_key` is set. This also allows `_handle_reuse_key_conflict` to be slightly simplified because we know `reuse_key` is set.

Reply:
Thanks. I think I got stuck on a certain idea about what `_update_renewal_params_from_key` and got lost. I think your feedback got me back on track :+1: .

>My initial impression is we shouldn't worry about the case of a manually modified renewal config file

Do you mean this scenario:

1. Renewal params have `reuse_key` set
2. User changes `key_type` in the renewal params
3. Certbot renews using the old key type

If so, I am inclined to agree that it's not worth it.

Reply:
Sorry for the confusion. I don't know about you, but I'm certainly finding grokking this code and all the different scenarios a bit tricky.

> Do you mean this scenario:
>
> 1. Renewal params have `reuse_key` set
> 2. User changes `key_type` in the renewal params
> 3. Certbot renews using the old key type
> 
> If so, I am inclined to agree that it's not worth it.

I meant a slightly different scenario. I meant:

1. Renewal params have `reuse_key` unset
2. User (perhaps accidentally) changes/deletes key type or size/curve in the renewal params
3. `certbot renew` will use renewal params changing the key's properties

In either your or my scenario though, there's a mismatch between renewal params and the current key which we could check for, but I don't think it's needed.

I just noticed this PR deviates from the current design doc though where I wrote:

> The raised error message, should tell people that the desired private key settings conflicts with --reuse-key. To change properties of the key, run Certbot with the certonly or run subcommand with the desired settings and without --reuse-key.

With this and https://github.com/certbot/certbot/issues/7711 in mind, I think it may be worth us looking at our current/desired behavior when reuse key is set and key type/size/curve is set on CLI and that differs from the existing key. Depending on our answers, I think it may also be worth thinking about `certonly/run --reuse-key` when a new lineage is created or if there are any differences if reuse_key is set on the CLI or not.

What do you think?

Reply:
> I meant a slightly different scenario. I meant:
> 
>   1. Renewal params have `reuse_key` unset
> 
>   2. User (perhaps accidentally) changes/deletes key type or size/curve in the renewal params
> 
>   3. `certbot renew` will use renewal params changing the key's properties

Isn't this the desired behavior? The design document describing that `certbot renew --key-type` (without `reuse_key`) *should* change the key type (or other key parameter). To me, it follows that modifying the renewal params file directly and then running `certbot renew` is the same operation.

>I just noticed this PR deviates from the current design doc though where I wrote
> > The raised error message, should tell people that the desired private key settings conflicts with --reuse-key. To change properties of the key, run Certbot with the certonly or run subcommand with the desired settings and without --reuse-key.

Do you mean that this PR just asks the user to add  `--cert-name` to their command (or alternatively, it will prompt them interactively)? If so: indeed.

Sorry to harp on about this again, but I dislike asking users to run `certonly/run` rather than `renew` to change a certificate, for having to remember and restate all the original parameters. The proposed documentation change in #9014 reflects the same viewpoint I have on this and https://github.com/certbot/certbot/issues/9009#issuecomment-906780606 is another one that comes to mind. I would like to understand whether and why we have different opinions on what's better here.  

>I think it may be worth us looking at our current/desired behavior when reuse key is set and key type/size/curve is set on CLI and that differs from the existing key

Yes, good point. Will look at this more closely.

Reply:
> > I meant a slightly different scenario. I meant:
> >
> >   1. Renewal params have `reuse_key` unset
> >
> >   2. User (perhaps accidentally) changes/deletes key type or size/curve in the renewal params
> >
> >   3. `certbot renew` will use renewal params changing the key's properties
>
> Isn't this the desired behavior?

I think that's reasonable and I think we're on the same page about not changing things here.

As for the rest of your message, I agree with you that using certonly/run to change renewal parameters isn't great. At the same time, I'm initially not sure I like the --cert-name approach either as I think the meaning of --reuse-key is still ambiguous.

I think us thinking carefully about our desired behavior around Certbot with reuse_key like I suggested in my last post should help here.